File: ninloader/go/collectors/aws/env.go
package aws

import (
	"fmt"
	"os"

	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

func init() {
	registry.Register("aws", "env", func() collector.Collector {
		return &EnvCollector{BaseCollector: collector.BaseCollector{Svc: "aws", Src: "env"}}
	})
}

// EnvCollector reads AWS credentials from environment variables.
type EnvCollector struct {
	collector.BaseCollector
}

func (c *EnvCollector) Service() string            { return c.Svc }
func (c *EnvCollector) Source() string              { return c.Src }
func (c *EnvCollector) StealthScore() int           { return 5 }
func (c *EnvCollector) Platforms() []string         { return nil }
func (c *EnvCollector) IsPlatformSupported() bool   { return true }

func (c *EnvCollector) Discover() []*types.DiscoveredToken {
	keyID := os.Getenv("AWS_ACCESS_KEY_ID")
	if keyID == "" {
		return nil
	}

	hint := keyID
	if len(hint) > 8 {
		hint = hint[:8]
	}

	return []*types.DiscoveredToken{
		{
			Service:      c.Svc,
			Source:       c.Src,
			AccountHint:  fmt.Sprintf("key_id=%s...", hint),
			StealthScore: c.StealthScore(),
			Details:      "environment variable",
		},
	}
}

func (c *EnvCollector) Collect() []*types.CollectedToken {
	keyID := os.Getenv("AWS_ACCESS_KEY_ID")
	secret := os.Getenv("AWS_SECRET_ACCESS_KEY")

	if keyID == "" || secret == "" {
		return nil
	}

	session := os.Getenv("AWS_SESSION_TOKEN")
	region := os.Getenv("AWS_DEFAULT_REGION")
	if region == "" {
		region = os.Getenv("AWS_REGION")
	}

	tok := types.NewCollectedToken(c.Svc, c.Src, c.StealthScore())
	tok.AccountID = keyID

	hint := keyID
	if len(hint) > 8 {
		hint = hint[:8]
	}
	tok.Username = fmt.Sprintf("env:%s...", hint)
	tok.AccessToken = types.Secure(keyID)
	tok.ClientSecret = types.Secure(secret)

	if session != "" {
		tok.Extra["session_token"] = session
	}
	if region != "" {
		tok.Extra["region"] = region
	}

	profile := os.Getenv("AWS_PROFILE")
	if profile != "" {
		tok.Extra["profile"] = profile
	}

	return []*types.CollectedToken{tok}
}

File: ninloader/python/ninloader/collectors/aws/env.py
"""AWS environment variable collector — reads AWS_ACCESS_KEY_ID, etc. from env."""

from __future__ import annotations

import os
from typing import List

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure


@CollectorRegistry.register
class AwsEnvCollector(BaseCollector):
    service = "aws"
    source = "env"
    stealth_score = 5

    def discover(self) -> List[DiscoveredToken]:
        key_id = os.environ.get("AWS_ACCESS_KEY_ID")
        if not key_id:
            return []

        return [DiscoveredToken(
            service=self.service,
            source=self.source,
            account_hint=f"key_id={key_id[:8]}...",
            stealth_score=self.stealth_score,
            details="environment variable",
        )]

    def collect(self) -> List[CollectedToken]:
        key_id = os.environ.get("AWS_ACCESS_KEY_ID")
        secret = os.environ.get("AWS_SECRET_ACCESS_KEY")

        if not key_id or not secret:
            return []

        session = os.environ.get("AWS_SESSION_TOKEN")
        region = os.environ.get("AWS_DEFAULT_REGION") or os.environ.get("AWS_REGION")

        extra = {}
        if session:
            extra["session_token"] = session
        if region:
            extra["region"] = region

        profile = os.environ.get("AWS_PROFILE")
        if profile:
            extra["profile"] = profile

        return [CollectedToken(
            service=self.service,
            source=self.source,
            stealth_score=self.stealth_score,
            account_id=key_id,
            username=f"env:{key_id[:8]}...",
            access_token=secure(key_id),
            client_secret=secure(secret),
            extra=extra,
        )]
